src/include/random.h

```diff
@@ -0,0 +1,397 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2017 SUSE LINUX GmbH
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+*/
+
+#ifndef CEPH_RANDOM_H
+#define CEPH_RANDOM_H 1
+
+#include <mutex>
+#include <random>
+
+#include "common/backport14.h"
+
+// Basic random number facility, adapted from N3551:
+namespace ceph {
+namespace util {
+
+inline namespace version_1_0 {
+
+namespace detail {
+
+template <typename T>
+T& engine()
+{
+  thread_local T rng_engine;
+  return rng_engine;
+}
+
+template <typename MutexT, typename EngineT>
+void randomize_rng(const int seed, MutexT& m, EngineT& e)
+{
+  std::lock_guard<MutexT> lg(m);
+  e.seed(seed);
+}
+
+template <typename MutexT, typename EngineT>
+void randomize_rng(MutexT& m, EngineT& e)
+{
+  thread_local std::random_device rd;
+ 
+  std::lock_guard<MutexT> lg(m);
+  e.seed(rd());
+}
+
+template <typename MutexT, 
+          typename EngineT = std::default_random_engine>
+void randomize_rng(MutexT& m)
+{
+  randomize_rng(m, detail::engine<EngineT>());
+}
+
+template <typename MutexT, 
+          typename EngineT = std::default_random_engine>
+void randomize_rng(const int n, MutexT& m)
+{
+  std::lock_guard<MutexT> lg(m);
+  detail::engine<EngineT>().seed(n);
+}
+
+template <typename EngineT = std::default_random_engine>
+void randomize_rng(const int n)
+{
+  detail::engine<EngineT>().seed(n);
+}
+
+template <typename EngineT = std::default_random_engine>
+void randomize_rng()
+{
+  thread_local std::random_device rd;
+  detail::engine<EngineT>().seed(rd());
+}
+} // namespace detail
+
+namespace detail {
```

Comment:
I see them as being generally unrelated to each other, there's no technical reason.